lib/spear/connection/request.ex

```diff
@@ -0,0 +1,194 @@
+defmodule Spear.Connection.Request do
+  @moduledoc false
+
+  # a struct representing a stream-able request
+
+  @type t :: %{
+          continuation: Enumerable.continuation(),
+          request_ref: Mint.Types.request_ref(),
+          buffer: binary(),
+          from: GenServer.from(),
+          response: Spear.Connection.Response.t(),
+          status: :streaming | :done
+        }
+
+  defstruct [:continuation, :buffer, :request_ref, :from, :response, :status]
+
+  def new(event_stream, request_ref, from) do
+    reducer = &reduce_with_suspend/2
+    stream = Stream.map(event_stream, &Spear.Request.to_wire_data/1)
+    continuation = &Enumerable.reduce(stream, &1, reducer)
+
+    %__MODULE__{
+      continuation: continuation,
+      buffer: <<>>,
+      request_ref: request_ref,
+      from: from,
+      response: %Spear.Connection.Response{},
+      status: :streaming
+    }
+  end
+
+  defp reduce_with_suspend(
+         {message, message_size},
+         {message_buffer, message_buffer_size, max_size}
+       )
+       when message_size + message_buffer_size > max_size do
+    {:suspend,
+     {[{message, message_size} | message_buffer], message_size + message_buffer_size, max_size}}
+  end
+
+  defp reduce_with_suspend(
+         {message, message_size},
+         {message_buffer, message_buffer_size, max_size}
+       ) do
+    {:cont,
+     {[{message, message_size} | message_buffer], message_size + message_buffer_size, max_size}}
+  end
+
+  @spec emit_messages(%Spear.Connection{}, %__MODULE__{}) ::
+          {:ok, %Spear.Connection{}} | {:error, %Spear.Connection{}, reason :: any()}
+  def emit_messages(state, %__MODULE__{buffer: <<>>, continuation: continuation} = request) do
+    smallest_window = get_smallest_window(state.conn, request.request_ref)
+
+    {:cont, {[], 0, smallest_window}}
+    |> continuation.()
+    |> handle_contination(state, request)
+  end
+
+  def emit_messages(
+        state,
+        %__MODULE__{buffer: buffer, continuation: continuation} = request
```

Comment:
if we have some bytes buffered in the our request in state

1. see if the buffer will fit in the new window
2. if it **does** fit in the new window
    - continue the stream using the current buffer as the first message in the reductions new accumulator
3. if it **does not** fit in the new window
    - take as much of the buffer as we can and say that the `rest` is our new buffer
    - go back to waiting for the server to tell us we have a new window

and this works! I have some events in my local EventStore now that are exactly 1048576B (~1MB) (where the typical HTTP2 window is 10-100KB)